File: easy-projects/easy-common/src/main/java/xyz/srclab/common/cache/ThreadLocalCache.java
package xyz.srclab.common.cache;

import xyz.srclab.annotation.Immutable;
import xyz.srclab.annotation.Nullable;

import java.time.Duration;
import java.util.Map;
import java.util.NoSuchElementException;
import java.util.function.Function;

/**
 * @author sunqian
 */
final class ThreadLocalCache<K, V> implements Cache<K, V> {

    private final ThreadLocal<Cache<K, V>> threadLocal;

    ThreadLocalCache(Cache<K, V> cache) {
        this.threadLocal = ThreadLocal.withInitial(() -> cache);
    }

    private Cache<K, V> getLocalCache() {
        return threadLocal.get();
    }

    @Override
    @Nullable
    public V load(K key, CacheLoader<? super K, ? extends V> loader) {
        return getLocalCache().load(key, loader);
    }

    @Override
    @Immutable
    public Map<K, @Nullable V> loadPresent(Iterable<? extends K> keys) {
        return getLocalCache().loadPresent(keys);
    }

    @Override
    @Immutable
    public Map<K, @Nullable V> loadAll(Iterable<? extends K> keys, CacheLoader<K, ? extends V> loader) {
        return getLocalCache().loadAll(keys, loader);
    }

    @Override
    public V loadNonNull(K key, CacheLoader<? super K, ? extends V> loader) throws NoSuchElementException {
        return getLocalCache().loadNonNull(key, loader);
    }

    @Override
    @Immutable
    public Map<K, V> loadPresentNonNull(Iterable<? extends K> keys) {
        return getLocalCache().loadPresentNonNull(keys);
    }

    @Override
    @Immutable
    public Map<K, V> loadAllNonNull(Iterable<? extends K> keys, CacheLoader<K, ? extends V> loader) throws NoSuchElementException {
        return getLocalCache().loadAllNonNull(keys, loader);
    }

    @Override
    public void putEntry(CacheEntry<? extends K, ? extends V> entry) {
        getLocalCache().putEntry(entry);
    }

    @Override
    public void putEntries(Iterable<? extends CacheEntry<? extends K, ? extends V>> cacheEntries) {
        getLocalCache().putEntries(cacheEntries);
    }

    @Override
    public void expire(K key, long seconds) {
        getLocalCache().expire(key, seconds);
    }

    @Override
    public void expire(K key, Duration duration) {
        getLocalCache().expire(key, duration);
    }

    @Override
    public void expireAll(Iterable<? extends K> keys, long seconds) {
        getLocalCache().expireAll(keys, seconds);
    }

    @Override
    public void expireAll(Iterable<? extends K> keys, Duration duration) {
        getLocalCache().expireAll(keys, duration);
    }

    @Override
    public void expireAll(Map<? extends K, Duration> expiryMap) {
        getLocalCache().expireAll(expiryMap);
    }

    @Override
    @Nullable
    public V get(K key) {
        return getLocalCache().get(key);
    }

    @Override
    @Nullable
    public V get(K key, Function<? super K, ? extends V> function) {
        return getLocalCache().get(key, function);
    }

    @Override
    @Immutable
    public Map<K, @Nullable V> getPresent(Iterable<? extends K> keys) {
        return getLocalCache().getPresent(keys);
    }

    @Override
    @Immutable
    public Map<K, @Nullable V> getAll(Iterable<? extends K> keys, Function<Iterable<? extends K>, Map<K, @Nullable V>> function) {
        return getLocalCache().getAll(keys, function);
    }

    @Override
    public V getNonNull(K key) throws NoSuchElementException {
        return getLocalCache().getNonNull(key);
    }

    @Override
    public V getNonNull(K key, Function<? super K, ? extends V> function) throws NoSuchElementException {
        return getLocalCache().getNonNull(key, function);
    }

    @Override
    @Immutable
    public Map<K, V> getPresentNonNull(Iterable<? extends K> keys) {
        return getLocalCache().getPresentNonNull(keys);
    }

    @Override
    @Immutable
    public Map<K, V> getAllNonNull(Iterable<? extends K> keys, Function<Iterable<? extends K>, Map<K, V>> function) throws NoSuchElementException {
        return getLocalCache().getAllNonNull(keys, function);
    }

    @Override
    public void put(K key, @Nullable V value) {
        getLocalCache().put(key, value);
    }

    @Override
    public void putAll(Map<K, @Nullable V> entries) {
        getLocalCache().putAll(entries);
    }

    @Override
    public void invalidate(K key) {
        getLocalCache().invalidate(key);
    }

    @Override
    public void invalidateAll(Iterable<? extends K> keys) {
        getLocalCache().invalidateAll(keys);
    }

    @Override
    public void invalidateALL() {
        getLocalCache().invalidateALL();
    }
}

File: boat-core/src/main/kotlin/xyz/srclab/common/cache/ThreadLocalCache.kt
package xyz.srclab.common.cache

import java.time.Duration

class ThreadLocalCache<K : Any, V>(cacheSupplier: () -> Cache<K, V>) : Cache<K, V> {

    private val threadLocal: ThreadLocal<Cache<K, V>> = ThreadLocal.withInitial(cacheSupplier)

    override fun get(key: K): V {
        return threadLocal.get().get(key)
    }

    override fun getOrNull(key: K): V? {
        return threadLocal.get().getOrNull(key)
    }

    override fun getOrElse(key: K, defaultValue: (K) -> V): V {
        return threadLocal.get().getOrElse(key, defaultValue)
    }

    override fun getPresent(keys: Iterable<K>): Map<K, V> {
        return threadLocal.get().getPresent(keys)
    }

    override fun getAll(keys: Iterable<K>, loader: (Iterable<K>) -> Map<K, V>): Map<K, V> {
        return threadLocal.get().getAll(keys, loader)
    }

    override fun getOrElse(key: K, defaultValue: V): V {
        return threadLocal.get().getOrElse(key, defaultValue)
    }

    override fun getOrLoad(key: K, loader: (K) -> V): V {
        return threadLocal.get().getOrLoad(key, loader)
    }

    override fun put(key: K, value: V) {
        threadLocal.get().put(key, value)
    }

    override fun put(key: K, value: V, expirySeconds: Long) {
        threadLocal.get().put(key, value, expirySeconds)
    }

    override fun put(key: K, value: V, expiry: Duration) {
        threadLocal.get().put(key, value, expiry)
    }

    override fun putAll(entries: Map<out K, V>) {
        threadLocal.get().putAll(entries)
    }

    override fun putAll(entries: Map<out K, V>, expirySeconds: Long) {
        threadLocal.get().putAll(entries, expirySeconds)
    }

    override fun putAll(entries: Map<out K, V>, expiry: Duration) {
        threadLocal.get().putAll(entries, expiry)
    }

    override fun expiry(key: K, expirySeconds: Long) {
        threadLocal.get().expiry(key, expirySeconds)
    }

    override fun expiry(key: K, expirySeconds: Duration) {
        threadLocal.get().expiry(key, expirySeconds)
    }

    override fun expiryAll(keys: Iterable<K>, expirySeconds: Long) {
        threadLocal.get().expiryAll(keys, expirySeconds)
    }

    override fun expiryAll(keys: Iterable<K>, expirySeconds: Duration) {
        threadLocal.get().expiryAll(keys, expirySeconds)
    }

    override fun invalidate(key: K) {
        threadLocal.get().invalidate(key)
    }

    override fun invalidateAll(keys: Iterable<K>) {
        threadLocal.get().invalidateAll(keys)
    }

    override fun invalidateAll() {
        threadLocal.get().invalidateAll()
    }

    override fun cleanUp() {
        threadLocal.get().cleanUp()
    }
}